src/gpu/gpu-cluster.ts

```diff
@@ -0,0 +1,213 @@
+import {
+    BUFFERUSAGE_COPY_DST,
+    BUFFERUSAGE_COPY_SRC,
+    SHADERLANGUAGE_WGSL,
+    SHADERSTAGE_COMPUTE,
+    BindGroupFormat,
+    BindStorageBufferFormat,
+    Compute,
+    FloatPacking,
+    Shader,
+    StorageBuffer,
+    WebgpuGraphicsDevice
+} from 'playcanvas/debug';
+
+import { DataTable } from '../data-table.js';
+import { GpuDevice } from './gpu-device.js';
+
+const clusterWgsl = (numColumns: number, numPoints: number, numCentroids: number) => {
+    return `
+
+enable f16;
+
+@group(0) @binding(0) var<storage, read> points: array<f16>;
+@group(0) @binding(1) var<storage, read> centroids: array<f16>;
+@group(0) @binding(2) var<storage, read_write> results: array<u32>;
+
+const numColumns = ${numColumns};
+const numPoints = ${numPoints};
+const numCentroids = ${numCentroids};
+
+const chunkSize = 128u; // must be a multiple of 64
+var<workgroup> sharedChunk: array<f16, numColumns * chunkSize>;
+
+// calculate the squared distance between the point and centroid
+fn calcDistanceSqr(point: array<f16, numColumns>, centroid: u32) -> f32 {
+    var result = 0.0;
+
+    var ci = centroid * numColumns;
+
+    for (var i = 0u; i < numColumns; i++) {
+        let v = f32(point[i] - sharedChunk[ci+i]);
+        result += v * v;
+    }
+
+    return result;
+}
+
+@compute @workgroup_size(64)
+fn main(
+    @builtin(local_invocation_index) local_id : u32,
+    @builtin(global_invocation_id) global_id: vec3u,
+    @builtin(num_workgroups) num_workgroups: vec3u
+) {
+    // calculate row index for this thread point
+    let pointIndex = global_id.x + global_id.y * num_workgroups.x * 64;
+
```

Comment:
do you need something like:
```
if (pointIndex < numPoints) { return; }
```

Reply:
You aren't able to return early when memory barriers are used - all threads in the workgroup must run the barrier code.

So I perform this out-of-bounds check before doing the actual work in a few places later on.

Reply:
ah right